mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,662 @@
+            cfg.system.hard_update,
+            hard_update,
+            soft_update,
+            *target_update_params,
+        )
+
+        # Repack params and opt_states.
+        next_params = DDQNParams(next_online_params, next_target_params)
+
+        return next_params, next_opt_state, q_loss_info
+
+    # TRAIN LEVEL 1
+    def train(train_state: TrainState, _: Any) -> Tuple[TrainState, Metrics]:
+        """Sample, train and repack."""
+
+        # unpack and get keys
+        buffer_state, params, opt_states, t_train, key = train_state
+        next_key, buff_key = jax.random.split(key, 2)
+
+        # sample
+        data = rb.sample(buffer_state, buff_key).experience
+
+        # learn
+        next_params, next_opt_states, q_loss_info = update_q(params, opt_states, data, t_train)
+
+        # Repack.
+        next_train_state = TrainState(
+            buffer_state, next_params, next_opt_states, t_train + 1, next_key
+        )
+
+        return next_train_state, q_loss_info
+
+    # INTERACT-TRAIN LOOP
+
+    scanned_interact = lambda state: lax.scan(
+        interaction_step, state, None, length=cfg.system.rollout_length
+    )
+    scanned_train = lambda state: lax.scan(train, state, None, length=cfg.system.epochs)
+
+    def update_step(
+        learner_state: LearnerState, _: Any
+    ) -> Tuple[LearnerState, Tuple[Metrics, Metrics]]:
+        """Interact, then learn."""
+
+        # unpack and get random keys
+        (
+            obs,
+            done,
+            hidden_state,
+            env_state,
+            time_steps,
+            train_steps,
+            opt_state,
+            buffer_state,
+            params,
+            key,
+        ) = learner_state
+        new_key, interact_key, train_key = jax.random.split(key, 3)
+
+        # Select actions, step env and store transitions
+        action_selection_state = ActionSelectionState(
+            params.online, hidden_state, time_steps, interact_key
+        )
+        interaction_state = InteractionState(
+            action_selection_state, env_state, buffer_state, obs, done
+        )
+        final_interaction_state, metrics = scanned_interact(interaction_state)
+
+        # Sample and learn
+        train_state = TrainState(
+            final_interaction_state.buffer_state, params, opt_state, train_steps, train_key
+        )
+        final_train_state, losses = scanned_train(train_state)
+
+        next_learner_state = LearnerState(
+            final_interaction_state.obs,
+            final_interaction_state.done,
+            final_interaction_state.action_selection_state.hidden_state,
+            final_interaction_state.env_state,
+            final_interaction_state.action_selection_state.time_steps,
+            final_train_state.train_steps,
+            final_train_state.opt_state,
+            final_interaction_state.buffer_state,
+            final_train_state.params,
+            new_key,
+        )
```

Comment:
It would be nice if you could compose your learner state of `TrainState`, `InteractionState` and a key